core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalLogWriter.java

```diff
@@ -149,24 +175,154 @@ public void close() throws IOException {
+   *    entry that has been persisted in UFS. Once we know the last valid journal entry in UFS,
+   *    we know the first journal entry that is not valid in UFS. Suppose it is Y.
+   * 3. Rename the incomplete journal file to X-Y. Future journal writes will write to a new file
+   *    named Y-0x7fffffffffffffff.
+   * 4. Check whether there is any missing journal entry between Y and the oldest entry in
+   *    mEntriesToFlush, say Z. If Z > Y, then it means journal entries in [Y, Z) are
+   *    missing, and Alluxio cannot recover. Otherwise, for each journal entry in
+   *    mEntriesToFlush, if its sequence number is newer than or equal to Y, retry writing it to
+   *    UFS by calling the {@code UfsJournalLogWriter#write} method.
+   */
+  private void maybeRecoverFromUfsFailures() throws IOException {
+    if (!mNeedsRecovery) {
+      return;
+    }
+    UfsJournalSnapshot snapshot = UfsJournalSnapshot.getSnapshot(mJournal);
+    long lastPersistSeq = -1;
+    UfsJournalFile currentLog = snapshot.getCurrentLog(mJournal);
+    if (currentLog != null) {
+      long startSeq = currentLog.getStart();
+      LOG.info("Recovering from previous UFS journal write failure."
+          + " Scanning for the last persisted journal entry.");
+      try (JournalReader reader = new UfsJournalReader(mJournal, startSeq, true)) {
+        JournalEntry entry;
+        while ((entry = reader.read()) != null) {
+          if (entry.getSequenceNumber() > lastPersistSeq) {
+            lastPersistSeq = entry.getSequenceNumber();
+          }
+        }
+      } catch (InvalidJournalEntryException e) {
+        LOG.info("Found last persisted journal entry with seq={}.", lastPersistSeq);
+      } catch (IOException e) {
+        throw e;
+      }
+      String src = currentLog.getLocation().toString();
+      if (lastPersistSeq >= 0) {
+        LOG.info("Last persisted journal entry sequence number in {} is {}",
+            src, lastPersistSeq);
+        String dst = UfsJournalFile
+            .encodeLogFileLocation(mJournal, currentLog.getStart(), lastPersistSeq + 1).toString();
+        if (mUfs.exists(dst)) {
+          LOG.warn("Deleting duplicate completed log {}.", dst);
+          // The dst can exist because of a master failure during commit. This can only happen
+          // when the primary master starts. We can delete either the src or dst. We delete dst and
+          // do rename again.
+          mUfs.deleteFile(dst);
+        }
+        LOG.info("Renaming the previous incomplete journal file from {} to {}.", src, dst);
+        mUfs.renameFile(src, dst);
+      } else {
+        LOG.info("No journal entry found in current journal file {}. Deleting it", src);
+        mUfs.deleteFile(src);
+      }
+    }
+    // Search for and scan the latest COMPLETE journal and find out the sequence number of the
+    // last persisted journal entry, in case no entry has been found in the INCOMPLETE journal.
+    if (lastPersistSeq < 0) {
```

Comment:
I'm afraid not. If we put this in the `else` block above, it will be executed only when the incomplete log exists (`currentLog != null`). However, we want to handle the case in which the incomplete log is totally lost.